### PR TITLE
Upgrade Rust to nightly-2024-01-10

### PR DESCRIPTION
**Pull request by orangecms**
Signed-off-by: Daniel Maslowski <info@orangecms.org>

Drop the Arm and PPC toolchains for now. They are not used and only waste space at this point.

### CODE REVIEW

File: src/lib/log/src/lib.rs [hunk 1]
```diff
@@ -132,6 +132,7 @@ pub fn print(args: fmt::Arguments) {
     LOGGER.lock().as_mut().unwrap().write_fmt(args).unwrap();
     #[cfg(not(feature = "mutex"))]
     unsafe {
+        // if let Some(l) = core::ptr::addr_of_mut!(LOGGER) { // not working
```
Comment (orangecms) [inline src/lib/log/src/lib.rs:135]: Help wanted: any idea how to properly write this? This here is what the Rust toolchain suggests when compiling, but it errors.

**Comment by orangecms** [comment]
let's see if upgrading dprint plugins fixes the recent issues

**Comment by orangecms** [comment]
> Error formatting /home/runner/work/oreboot/oreboot/src/soc/src/amd/common/mp/mpmailbox.rs. Message: Child process exited with code 1: error: the 'rustfmt' binary, normally provided by the 'rustfmt' component, is not applicable to the 'nightly-2024-01-10-x86_64-unknown-linux-gnu' toolchain

fine, I'll try 2024-01-09......

**Comment by orangecms** [comment]
back to 2024-01-10 - and remove the quickformat check; it just doesn't work

we have the check in place with our regular setup, and that actually works, just takes a little longer

**Comment by orangecms** [comment]
Note: generator has been renamed into coroutine in Rust

https://github.com/rust-lang/rust/commit/e96ce20b34789d29e925425da6cf138927b80a79

File: src/lib/log/src/lib.rs [hunk 2]
```diff
@@ -132,6 +132,7 @@ pub fn print(args: fmt::Arguments) {
...
         if let Some(l) = &mut LOGGER {
```
Comment (rmsyn) [inline src/lib/log/src/lib.rs:136]:
```suggestion
        if let Some(l) = LOGGER.as_mut() {
```

File: src/mainboard/starfive/visionfive2/bt0/src/main.rs [hunk 3]
```diff
@@ -9,7 +9,12 @@ extern crate log;
 extern crate layoutflash;
 use layoutflash::areas::{find_fdt, FdtIterator};
 
-use core::{arch::asm, intrinsics::transmute, panic::PanicInfo, ptr};
+use core::{
+    arch::asm,
+    intrinsics::transmute,
+    panic::PanicInfo,
+    ptr::{self, addr_of_mut},
```
Comment (rmsyn) [inline src/mainboard/starfive/visionfive2/bt0/src/main.rs:16]:
```suggestion
    ptr::{self, addr_of, addr_of_mut},
```

File: src/mainboard/starfive/visionfive2/bt0/src/main.rs [hunk 4]
```diff
@@ -127,11 +132,17 @@ pub unsafe extern "C" fn reset() {
         static _sidata: u8;
     }
 
-    let count = &_ebss as *const u8 as usize - &_sbss as *const u8 as usize;
-    ptr::write_bytes(&mut _sbss as *mut u8, 0, count);
-
-    let count = &_edata as *const u8 as usize - &_sdata as *const u8 as usize;
-    ptr::copy_nonoverlapping(&_sidata as *const u8, &mut _sdata as *mut u8, count);
+    let count =
+        addr_of_mut!(_ebss) as *const u8 as usize - addr_of_mut!(_sbss) as *const u8 as usize;
```
Comment (rmsyn) [inline src/mainboard/starfive/visionfive2/bt0/src/main.rs:136]:
```suggestion
        addr_of!(_ebss) as usize - addr_of!(_sbss) as usize;
```

File: src/mainboard/starfive/visionfive2/bt0/src/main.rs [hunk 5]
```diff
@@ -127,11 +132,17 @@ pub unsafe extern "C" fn reset() {
...
+    ptr::write_bytes(addr_of_mut!(_sbss) as *mut u8, 0, count);
+
+    let count =
+        addr_of_mut!(_edata) as *const u8 as usize - addr_of_mut!(_sdata) as *const u8 as usize;
```
Comment (rmsyn) [inline src/mainboard/starfive/visionfive2/bt0/src/main.rs:140]:
```suggestion
        addr_of!(_edata) as usize - addr_of!(_sdata) as usize;
```

File: src/mainboard/starfive/visionfive2/bt0/src/main.rs [hunk 6]
```diff
@@ -127,11 +132,17 @@ pub unsafe extern "C" fn reset() {
...
+    ptr::copy_nonoverlapping(
+        &_sidata as *const u8,
```
Comment (rmsyn) [inline src/mainboard/starfive/visionfive2/bt0/src/main.rs:142]:
```suggestion
        addr_of!(_sidata),
```

File: src/mainboard/starfive/visionfive2/bt0/src/main.rs [hunk 7]
```diff
@@ -127,11 +132,17 @@ pub unsafe extern "C" fn reset() {
...
+        addr_of_mut!(_sdata) as *mut u8,
```
Comment (rmsyn) [inline src/mainboard/starfive/visionfive2/bt0/src/main.rs:143]:
```suggestion
        addr_of_mut!(_sdata),
```

File: src/mainboard/starfive/visionfive2/bt0/src/main.rs [hunk 8]
```diff
@@ -127,11 +132,17 @@ pub unsafe extern "C" fn reset() {
...
     }
 
-    let count = &_ebss as *const u8 as usize - &_sbss as *const u8 as usize;
-    ptr::write_bytes(&mut _sbss as *mut u8, 0, count);
-
-    let count = &_edata as *const u8 as usize - &_sdata as *const u8 as usize;
-    ptr::copy_nonoverlapping(&_sidata as *const u8, &mut _sdata as *mut u8, count);
+    let count =
+        addr_of_mut!(_ebss) as *const u8 as usize - addr_of_mut!(_sbss) as *const u8 as usize;
+    ptr::write_bytes(addr_of_mut!(_sbss) as *mut u8, 0, count);
```
Comment (rmsyn) [inline src/mainboard/starfive/visionfive2/bt0/src/main.rs:137]:
```suggestion
    ptr::write_bytes(addr_of_mut!(_sbss), 0, count);
```

**Review by rmsyn — CHANGES_REQUESTED** [review]
Few small suggestions to remove unnecessary casts.

Added a suggestion for the help requested.

File: src/mainboard/starfive/visionfive2/bt0/src/main.rs (hunk 5 above)
Comment (orangecms) [reply src/mainboard/starfive/visionfive2/bt0/src/main.rs:140]: That is so cool, Rust is really getting more and more elegant all the time :) thank you!

File: src/lib/log/src/lib.rs (hunk 2 above)
Comment (orangecms) [reply src/lib/log/src/lib.rs]: Done, thank you so much!

**Review by rminnich — APPROVED** [review]

**Pull request merged**